pigosat.go

```diff
@@ -312,3 +324,87 @@ func (p *Pigosat) BlockSolution(solution Solution) error {
 	p.blocksol(solution)
 	return nil
 }
+
+// Write the clauses that were used in deriving the empty clause to a file
+// in DIMACS format.
+//
+// Requires that Pigosat was created with EnableTrace == true.
+func (p *Pigosat) WriteClausalCore(f io.Writer) error {
+	defer p.ready(true)()
+	if Status(C.picosat_res(p.p)) != Unsatisfiable {
+		return errors.New("expected to be in Unsatisfiable state")
+	}
+
+	return cFileWriterWrapper(f, func(cfile *C.FILE) error {
+		// void picosat_write_clausal_core (PicoSAT *, FILE * core_file);
+		_, err := C.picosat_write_clausal_core(p.p, cfile)
+		return err
+	})
+}
+
+// Write a compact proof trace in TraceCheck format to a file.
+//
+// Requires that Pigosat was created with EnableTrace == true.
+func (p *Pigosat) WriteCompactTrace(f io.Writer) error {
+	defer p.ready(true)()
+	if Status(C.picosat_res(p.p)) != Unsatisfiable {
+		return errors.New("expected to be in Unsatisfiable state")
+	}
+
+	return cFileWriterWrapper(f, func(cfile *C.FILE) error {
+		// void picosat_write_compact_trace (PicoSAT *, FILE * trace_file);
+		_, err := C.picosat_write_compact_trace(p.p, cfile)
+		return err
+	})
+}
+
+// Write an extended proof trace in TraceCheck format to a file.
+//
+// Requires that Pigosat was created with EnableTrace == true.
+func (p *Pigosat) WriteExtendedTrace(f io.Writer) error {
+	defer p.ready(true)()
+	if Status(C.picosat_res(p.p)) != Unsatisfiable {
+		return errors.New("expected to be in Unsatisfiable state")
+	}
+
+	return cFileWriterWrapper(f, func(cfile *C.FILE) error {
+		// void picosat_write_extended_trace (PicoSAT *, FILE * trace_file);
+		_, err := C.picosat_write_extended_trace(p.p, cfile)
+		return err
+	})
+}
+
+// A wrapper to take an io.Writer interface and feed it
+// the output from picosat functions that specifically
+// need a *C.FILE
+func cFileWriterWrapper(w io.Writer, writeFn func(*C.FILE) error) (err error) {
```

Comment:
1. I got rid of the goroutine and error channel because I felt like it added complexity to a part of the code that I suspect won't ever be the speed bottleneck. Let me know if you disagree.
2. I tried to catch errors that come out of all the files we need to close here.
3. `wp` and `cfile` now both close *after* the read from `rp`. From what I can tell reading the [documentation](https://golang.org/pkg/os/#Pipe), there's no need to flush, sync, or close the write end of the pipe before reading from the read end of the pipe. I'm not sure we need to flush or close `cfile` for all of `cfile`'s content to make it into `wp`, but I suspect we don't. Thoughts?